### ISSUE TITLE
Tensorflow combining with PySpark pandas_udf yields "triggered tf.function retracing".

### ISSUE THREAD

**Issue by tungnat97** [tensorflow/tensorflow#44257]
Here is a snippet of my code combining Tensorflow with Spark for online prediction:
```
@pandas_udf(StringType())
def online_predict(values: pd.Series) -> pd.Series:
    pred = Model.from_config(bc_config.value)
    pred.set_weights(bc_weights.value)
    ds = tf.data.Dataset.from_tensor_slices(values)
    ds = ds.map(preprocessing).batch(batch_size)
    res = pred.predict_step(ds)
    res = tf.norm(res, axis=1)
    # res = tf.greater(res, 5.0)
    res = tf.strings.as_string(res).numpy()
    return pd.Series(res)


spark = SparkSession.builder.appName(
    'spark_tf').master("local[*]").getOrCreate()
weights = np.load('./ext/weights.npy', allow_pickle=True)
config = np.load('./ext/config.npy', allow_pickle=True).item()
bc_weights = spark.sparkContext.broadcast(weights)
bc_config = spark.sparkContext.broadcast(config)

stream = spark.readStream.format('kafka') \
    .option('kafka.bootstrap.servers', 'localhost:9092') \
    .option('subscribe', 'network') \
    .load()

stream = stream.select(online_predict(col('value')).alias('value'))

x = stream.writeStream \
    .format('kafka') \
    .option("kafka.bootstrap.servers", 'localhost:9092') \
    .option('topic', 'dlpred') \
    .option('checkpointLocation', './kafka_checkpoint') \
    .start()

x.awaitTermination()
```

Running the code yields this warning when data come repeatedly from Kafka topic, I assume this is due to the repeat of creating dataset:

> WARNING:tensorflow:7 out of the last 7 calls to <function Model.make_predict_function.<locals>.predict_function at 0x7f8ca6abf598> triggered tf.function retracing. Tracing is expensive and the excessive number of tracings could be due to (1) creating @tf.function repeatedly in a loop, (2) passing tensors with different shapes, (3) passing Python objects instead of tensors. For (1), please define your @tf.function outside of the loop. For (2), @tf.function has experimental_relax_shapes=True option that relaxes argument shapes that can avoid unnecessary retracing. For (3), please refer to https://www.tensorflow.org/tutorials/customization/performance#python_or_tensor_args and https://www.tensorflow.org/api_docs/python/tf/function for  more details.

**Comment by ravikyram** [comment]
@tungnat97 

Request you to share complete code snippet with supporting files to reproduce the issue in our environment.It helps us in localizing the issue faster.

If you want to suppress the warnings please use the below code before importing tensorflow.

```
import os
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '2'
import tensorflow as tf
```
Thanks!

**Comment by tungnat97** [comment]
As for now I can't provide the full code as the code uses other libraries rather than Tensorflow and also interacts with a local server. I am also not able to create an alternative code right now, but the problem seems to come from the `predict` function in the code snippet I posted earlier, as the problem was gone after I changed it to `predict_step` inside a loop of the dataset. I'm not dealing with sequence data (tabular data) so it's not due to the variable length problem. I will try to come up with an alternative code asap.
